File: codes/julia/ch18/eigen_benchmark_finite.jl
# eigen_benchmark_finite.jl
#
# Chapter 18: Linear Spectral Eigenproblems
# Computational Etude 18.3: Finite-interval benchmark and the N/2 rule.
#
# Repeats the Dirichlet Laplacian at N = 16, 32, 64 and plots the
# absolute eigenvalue error on a semilog scale vs. mode number j.
# Reproduces Boyd (2000) Figs. 7.1 and 7.3.
#
# Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
# Part of "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using Printf

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

_json_val(x::Bool)          = x ? "true" : "false"
_json_val(x::Integer)       = string(x)
_json_val(x::Real)          = isfinite(x) ? string(Float64(x)) : (isnan(x) ? "NaN" : (x > 0 ? "Infinity" : "-Infinity"))
_json_val(x::AbstractString)= "\"" * replace(String(x), "\\" => "\\\\", "\"" => "\\\"") * "\""
_json_val(v::AbstractVector)= "[" * join(_json_val.(v), ", ") * "]"
function _json_val(d::AbstractDict)
    pairs = String[]
    for (k, v) in d; push!(pairs, _json_val(string(k)) * ": " * _json_val(v)); end
    return "{" * join(pairs, ", ") * "}"
end
write_json(path, x) = open(io -> print(io, _json_val(x)), path, "w")

set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 11,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 9,),
))

const NAVY  = colorant"#142D6E"
const CORAL = colorant"#E74C3C"
const TEAL  = colorant"#16A085"

const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch18", "julia")
mkpath(OUTPUT_DIR)

function solve_dirichlet(N::Int)
    D, _ = cheb_matrix(N)
    D2 = D * D
    A = -D2[2:N, 2:N]
    lam = real.(eigvals(A))
    return sort(lam)
end

function count_good(err::Vector{<:Real}, tol::Real = 1e-2)
    for (k, e) in enumerate(err)
        if e >= tol
            return k - 1
        end
    end
    return length(err)
end

function make_figure(; Ns = (16, 32, 64), dump_path = nothing)
    colours = [NAVY, CORAL, TEAL]
    markers = [:circle, :rect, :utriangle]
    counts = Dict{Int,Int}()
    spectra = Dict{Int,Vector{Float64}}()

    fig = Figure(size = (760, 460))
    ax  = Axis(fig[1, 1];
        xlabel = "mode number j",
        ylabel = "|λ_num − λ_exact|",
        title  = "Dirichlet Laplacian: errors at three resolutions",
        yscale = log10,
        limits = ((0, maximum(Ns) + 2), (1e-17, 1e6)))

    for (N, c, m) in zip(Ns, colours, markers)
        lam = solve_dirichlet(N)
        j = collect(1:length(lam))
        lam_exact = (j .* π ./ 2) .^ 2
        err = abs.(lam .- lam_exact)
        spectra[N] = lam
        counts[N] = count_good(err, 1e-2)
        err_plot = max.(err, 1e-17)
        scatterlines!(ax, j, err_plot;
                      color = c, markercolor = :white,
                      strokecolor = c, strokewidth = 1.0,
                      linewidth = 0.6, marker = m, markersize = 8,
                      label = "N = $N")
        vlines!(ax, [N / 2]; color = c, linestyle = :dot,
                linewidth = 0.7, alpha = 0.5)
    end

    hlines!(ax, [1e-2]; color = :black, linestyle = :dash,
            linewidth = 0.6, alpha = 0.5)
    text!(ax, 1.5, 2e-2; text = "0.01 tolerance",
          fontsize = 8, color = (:black, 0.7))
    axislegend(ax, position = :rb, labelsize = 9, patchsize = (12, 12))

    save(joinpath(OUTPUT_DIR, "eigen_benchmark_finite.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "eigen_benchmark_finite.png"), fig, px_per_unit = 4)

    r = Dict(
        "Ns"           => collect(Ns),
        "spectra"      => Dict(string(k) => v for (k, v) in spectra),
        "good_counts"  => Dict(string(k) => v for (k, v) in counts),
    )
    dump_path !== nothing && write_json(dump_path, r)
    return r, counts, spectra
end

function parse_args()
    dump_path = nothing; i = 1
    while i <= length(ARGS)
        if ARGS[i] == "--dump" && i < length(ARGS); dump_path = ARGS[i+1]; i += 2
        else i += 1 end
    end
    return dump_path
end

if abspath(PROGRAM_FILE) == @__FILE__
    dp = parse_args()
    _, counts, _ = make_figure(; dump_path = dp)
    println("[Etude 18.3]")
    for N in sort(collect(keys(counts)))
        @printf("  N = %2d:  %d eigenvalues with |err| < 1e-2   (heuristic N/2 = %d)\n",
                N, counts[N], N ÷ 2)
    end
    println("  figure: ", joinpath(OUTPUT_DIR, "eigen_benchmark_finite.pdf"))
    dp !== nothing && println("  dumped: ", dp)
end

File: codes/python/ch18/eigen_benchmark_finite.py
#!/usr/bin/env python3
"""
eigen_benchmark_finite.py

Chapter 18: Linear Spectral Eigenproblems
Computational Etude 18.3: Finite-interval benchmark and the N/2 rule.

Repeats the Dirichlet Laplacian calculation at N = 16, 32, 64 and plots
the absolute error |lambda_j^num - lambda_j^exact| on a semilog scale
against mode number j. Reproduces Boyd (2000) Fig 7.1 and Fig 7.3.

At each N, the lowest N/2 eigenvalues are accurate to within a few
percent (Boyd's Rule-of-Thumb 8); beyond that threshold the error
climbs exponentially until it saturates near the eigenvalue magnitude
itself (random, meaningless numbers).

Interpretation of the figure:
  - The flat roundoff floor at 1e-13 for low modes is the combined effect
    of floating-point accumulation in eig() and the sigma-scaled
    condition number of the interior Laplacian.
  - The steep rise starts near j ~ N/2 (vertical dividing lines in the
    plot) and levels off at O(lambda_j) for the unresolved modes.
  - Doubling N shifts the "cliff" rightward by roughly N, buying an
    equal number of additional trusted eigenvalues --- EXACTLY the
    behaviour Boyd's heuristic predicts.

Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
Part of "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

SCRIPT_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(SCRIPT_DIR.parent / "ch07"))
from cheb_matrix import cheb_matrix  # noqa: E402

plt.rcParams.update({
    "font.family": "serif",
    "font.serif": ["CMU Serif", "DejaVu Serif"],
    "mathtext.fontset": "cm",
    "font.size": 11,
    "axes.linewidth": 0.8,
    "xtick.major.width": 0.8,
    "ytick.major.width": 0.8,
    "figure.dpi": 150,
    "savefig.dpi": 300,
})

NAVY, SKY, CORAL = "#142D6E", "#7896D2", "#E74C3C"
TEAL, PURPLE, ORANGE = "#16A085", "#8E44AD", "#E67E22"

OUTPUT_DIR = SCRIPT_DIR.parent.parent.parent / "textbook" / "figures" / "ch18" / "python"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


def solve_dirichlet(N: int) -> np.ndarray:
    """Return sorted interior-block spectrum of u_xx + lam u = 0, u(+-1) = 0."""
    D, _ = cheb_matrix(N)
    D2 = D @ D
    A = -D2[1:N, 1:N]
    lam = np.linalg.eigvals(A).real
    return np.sort(lam)


def count_good(err: np.ndarray, tol: float = 1e-2) -> int:
    """Return the LARGEST j such that every error up to j is below tol."""
    below = err < tol
    if not below.any():
        return 0
    # first failure index
    failures = np.where(~below)[0]
    return failures[0] if failures.size > 0 else err.size


def make_figure(Ns=(16, 32, 64)):
    colours = [NAVY, CORAL, TEAL]
    markers = ["o", "s", "^"]
    counts = {}

    fig, ax = plt.subplots(1, 1, figsize=(7.6, 4.6))

    for N, c, m in zip(Ns, colours, markers):
        lam = solve_dirichlet(N)
        j = np.arange(1, lam.size + 1)
        lam_exact = (j * np.pi / 2) ** 2
        err = np.abs(lam - lam_exact)
        err_plot = np.maximum(err, 1e-17)
        ax.semilogy(j, err_plot, marker=m, linestyle="-", color=c,
                    markerfacecolor="white", markeredgewidth=1.0,
                    markersize=5, linewidth=0.6,
                    label=f"$N = {N}$")
        # vertical cue at N/2
        ax.axvline(N / 2, color=c, linestyle=":", linewidth=1.4, alpha=0.85)
        counts[N] = count_good(err, tol=1e-2)

    ax.axhline(1e-2, color="k", linestyle="--", linewidth=0.6, alpha=0.5)
    ax.text(1.5, 2e-2, "0.01 tolerance", fontsize=8, color="k", alpha=0.7)
    ax.set_xlabel("mode number $j$")
    ax.set_ylabel(r"$|\lambda^{\mathrm{num}}_j - \lambda^{\mathrm{exact}}_j|$")
    ax.set_title("Dirichlet Laplacian: errors at three resolutions",
                 fontsize=11)
    ax.set_xlim(0, max(Ns) + 2)
    ax.set_ylim(1e-17, 1e6)
    ax.grid(True, which="both", alpha=0.25, linewidth=0.4)
    ax.legend(loc="lower right", fontsize=9, framealpha=0.9)

    fig.tight_layout()
    fig.savefig(OUTPUT_DIR / "eigen_benchmark_finite.pdf", bbox_inches="tight", pad_inches=0.05)
    fig.savefig(OUTPUT_DIR / "eigen_benchmark_finite.png", bbox_inches="tight", pad_inches=0.05)
    plt.close(fig)

    spectra = {int(N): solve_dirichlet(N).tolist() for N in Ns}
    return {
        "Ns":            list(Ns),
        "spectra":       spectra,
        "good_counts":   {int(N): int(k) for N, k in counts.items()},
    }


def main():
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    p.add_argument("--dump", type=str, default=None)
    args = p.parse_args()

    r = make_figure()
    print(f"[Etude 18.3]")
    for N in r["Ns"]:
        print(f"  N = {N:2d}:  {r['good_counts'][N]} eigenvalues with |err| < 1e-2   (heuristic N/2 = {N//2})")
    print(f"  figure: {OUTPUT_DIR / 'eigen_benchmark_finite.pdf'}")
    if args.dump:
        Path(args.dump).write_text(json.dumps(r, indent=2))
        print(f"  dumped: {args.dump}")


if __name__ == "__main__":
    main()
